File: packages/edge/src/routes/health.ts
import { Router, Request, Response } from 'express';
import { sdkManager } from '../services/sdkManager';
import { tokenMirrorService } from '../services/tokenMirrorService';

const router = Router();

/**
 * Health check endpoint
 * GET /health
 */
router.get('/', (req: Request, res: Response) => {
  const sdk = sdkManager.getSDK();
  const isSdkReady = sdk !== null;
  const isTokenMirrorReady = tokenMirrorService.isInitialized();
  const isReady = isSdkReady && isTokenMirrorReady;

  res.status(isReady ? 200 : 503).json({
    status: isReady ? 'healthy' : 'initializing',
    timestamp: new Date().toISOString(),
    version: '1.0.0',
    sdk: isSdkReady ? 'ready' : 'initializing',
    tokenMirror: isTokenMirrorReady ? 'ready' : 'initializing',
    tokenCount: tokenMirrorService.getTokenCount(),
  });
});

/**
 * Readiness check endpoint
 * GET /health/ready
 */
router.get('/ready', (req: Request, res: Response) => {
  const sdk = sdkManager.getSDK();
  const isReady = sdk !== null;

  if (isReady) {
    res.status(200).json({
      status: 'ready',
      timestamp: new Date().toISOString(),
    });
  } else {
    res.status(503).json({
      status: 'not_ready',
      message: 'SDK not initialized',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * Liveness check endpoint
 * GET /health/live
 */
router.get('/live', (req: Request, res: Response) => {
  res.status(200).json({
    status: 'alive',
    timestamp: new Date().toISOString(),
  });
});

export default router;

File: packages/edge-rust/src/routes/health.rs
// Gatrix Edge Rust - Health Routes

use actix_web::{web, HttpResponse, Responder};
use crate::app_state::AppState;

/// GET /health
pub async fn health(data: web::Data<AppState>) -> impl Responder {
  let sdk = data.sdk.read().await;
  let is_sdk_ready = sdk.is_initialized();
  let is_token_mirror_ready = data.token_mirror.is_initialized().await;
  let is_ready = is_sdk_ready && is_token_mirror_ready;

  let status_code = if is_ready { 200 } else { 503 };
  let token_count = data.token_mirror.get_token_count().await;

  HttpResponse::build(actix_web::http::StatusCode::from_u16(status_code).unwrap())
    .json(serde_json::json!({
      "status": if is_ready { "healthy" } else { "initializing" },
      "timestamp": chrono::Utc::now().to_rfc3339(),
      "version": env!("CARGO_PKG_VERSION"),
      "runtime": "rust",
      "sdk": if is_sdk_ready { "ready" } else { "initializing" },
      "tokenMirror": if is_token_mirror_ready { "ready" } else { "initializing" },
      "tokenCount": token_count,
    }))
}

/// GET /health/ready
pub async fn health_ready(data: web::Data<AppState>) -> impl Responder {
  let sdk = data.sdk.read().await;
  let is_ready = sdk.is_initialized();

  if is_ready {
    HttpResponse::Ok().json(serde_json::json!({
      "status": "ready",
      "timestamp": chrono::Utc::now().to_rfc3339(),
    }))
  } else {
    HttpResponse::ServiceUnavailable().json(serde_json::json!({
      "status": "not_ready",
      "message": "SDK not initialized",
      "timestamp": chrono::Utc::now().to_rfc3339(),
    }))
  }
}

/// GET /health/live
pub async fn health_live() -> impl Responder {
  HttpResponse::Ok().json(serde_json::json!({
    "status": "alive",
    "timestamp": chrono::Utc::now().to_rfc3339(),
  }))
}

/// Configure health routes
pub fn configure(cfg: &mut web::ServiceConfig) {
  cfg.service(
    web::scope("/health")
      .route("", web::get().to(health))
      .route("/", web::get().to(health))
      .route("/ready", web::get().to(health_ready))
      .route("/live", web::get().to(health_live)),
  );
}
